lib/Conversion/GPUToROCM/ConvertKernelFuncToHSACO.cpp

```diff
@@ -0,0 +1,401 @@
+      // "getROCDLPaths" is stable/accurate.
+      // return failure();
+      continue;
+    }
+
+    llvm::SMDiagnostic diagnostic;
+    std::unique_ptr<llvm::Module> bitcodeModule(
+        llvm::parseIRFile(llvm::StringRef(filename.data(), filename.size()),
+                          diagnostic, llvmModule.getContext()));
+
+    if (bitcodeModule == nullptr) {
+      MLIRContext *mlirContext = function.getContext();
+      auto parseErrorLocation = mlir::FileLineColLoc::get(
+          diagnostic.getFilename().str(), diagnostic.getLineNo(),
+          diagnostic.getColumnNo(), mlirContext);
+      mlir::emitError(parseErrorLocation, diagnostic.getMessage().str());
+      function.emitError("Error parsing ROCDL bitcode module from " + filename);
+      return failure();
+    }
+
+    if (linker.linkInModule(
+            std::move(bitcodeModule), llvm::Linker::Flags::LinkOnlyNeeded,
+            [](llvm::Module &M, const llvm::StringSet<> &GVS) {
+              internalizeModule(M, [&M, &GVS](const llvm::GlobalValue &GV) {
+                return !GV.hasName() || (GVS.count(GV.getName()) == 0);
+              });
+            })) {
+      function.emitError("Error linking ROCDL bitcode module from " + filename);
+      return failure();
+    }
+  }
+
+  return success();
+}
+
+// Returns whether the module uses any ROCDL bitcode functions. This function
+// may have false positives
+static bool couldNeedDeviceBitcode(const llvm::Module &llvmModule) {
+  for (const llvm::Function &llvmFunction : llvmModule.functions()) {
+    // This is a conservative approximation
+    //  - not all such functions are in ROCm-Device-Libs.
+    if (!llvmFunction.isIntrinsic() && llvmFunction.isDeclaration())
+      return true;
+  }
+  return false;
+}
+
+// Links ROCm-Device-Libs into the given module if the module needs it.
+static LogicalResult linkROCDLIfNecessary(llvm::Module &llvmModule,
+                                          FuncOp function,
+                                          rocm::AMDGPUVersion amdgpuVersion,
+                                          llvm::StringRef rocdlDir) {
+
+  if (!couldNeedDeviceBitcode(llvmModule))
+    return success();
+
+  return linkWithBitcodeModules(llvmModule, function,
+                                getROCDLPaths(amdgpuVersion, rocdlDir));
+}
+
+// Emits the given module to HSA Code Object. targetMachine is an initialized
+// TargetMachine for the AMDGPU target.
+OwnedHSACO
+GpuKernelToHSACOPass::emitModuleToHSACO(llvm::Module &llvmModule,
+                                        FuncOp function,
+                                        llvm::TargetMachine &targetMachine) {
+  llvm::SmallString<128> tempdirName;
+  if (llvm::sys::fs::createUniqueDirectory("/tmp/amdgpu_mlir", tempdirName)) {
+    llvm::errs() << "Failed to create tempdir for generating HSACO\n";
+    return {};
+  }
+
+  // prepare filenames for all stages of compilation:
+  // IR, ISA, binary ISA, and HSACO
```

Comment:
Please add a TODO and file a bug to replace this with using lld as a library for the link.